docs/developer-docs/security/general-security-best-practices.md

```diff
@@ -107,3 +107,25 @@ For example, we have seen issues where the public key to verify certification wa
 #### Recommendation
 
 Avoid test and dev code in production code whenever possible.
+
+### Data confidentiality on the Internet Computer
+
+#### Security concern
+
+When storing data on the Internet Computer, there are two levels of data access.
+
+1. Nodes are able to read all data that is stored on a subnet. This includes all messages sent to or from a canister, along with all data stored in a canister. This means a node could extract all data available to a canister. This will change with the implementation of TEE-based security for nodes.
+
+2. End user clients can only access whatever data that nodes and canisters have made available to them. If a node does not misbehave and leak data, clients can only read the responses to ingress messages and queries that they have sent. The canister decides what data is exposed to the client. 
+
+Partial information on data that is stored in the subnet state tree will always leak. Therefore, data with a low-entropy value may entirely leak and be fully exposed, such as a Boolean value that can only be either "True" or "False". Leakage on data with a high-entropy is negligible. 
+
+There are two types of user-related data that may be stored in the subnet state tree. The first is when a user sends an ingress message to a canister, the message and the response are both stored in the subnet state tree to be retrieved securely by the client. The ingress message should contain a high-entropy nonce that is implemented by the agent and typically not exposed to the user. The message response is determined by the canister and may not contain a high-entropy value. If the canister response consists of a low-entropy value, then the data may be leaked to users other than the ingress message sender. 
+
+The second type of user-related data is certified variables maintained by a canister that are also exposed through the subnet state tree. If a canister places low-entropy data into the state tree, then the data may leak to users who should not have access to that piece of data. 
+
+#### Recommendation 
```

Comment:
minor thing: would it make sense to also link to the Interface spec section that @mraszyk is writing [here](https://github.com/dfinity/interface-spec/pull/209/files)?

Reply:
Added a link to the IC spec